### ISSUE TITLE
 JSON.ToObject<TValue>(stringValue)

### ISSUE THREAD

**Issue by BogdanovAnton** [mgholam/fastJSON#39]
JSON.ToObject<TValue>(stringValue) manipulating with global settings!
Here stack:
   JSON.ToObject< TValue >(stringValue)
 ->
    public static T ToObject< T >(string json)
    {
      return new deserializer(**JSON.Parameters**).ToObject< T >(json);  // Oops...
    }
 ->
    public deserializer(JSONParameters param)
    {
      **this._params = param;** // O-o !
    }
 ->
    public T ToObject< T >(string json)
    {
      Type type = typeof (T);
      object obj = this.ToObject(json, type);
      if (!type.IsArray)
        return (T) obj;
      if ((obj as ICollection).Count == 0)
        return (T) (object) Array.CreateInstance(type.GetElementType(), 0);
      return (T) obj;
    }
->
    public object ToObject(string json, Type type)
    {
      this._params.FixValues();
      Type type1 = (Type) null;
      if (type != (Type) null && type.IsGenericType)
        type1 = Reflection.Instance.GetGenericTypeDefinition(type);
      if (type1 == typeof (Dictionary< , >) || type1 == typeof (List< >))
        **this._params.UsingGlobalTypes = false;**  // Bug here !    <----
      this._usingglobals = this._params.UsingGlobalTypes;
      object parse = new JsonParser(json).Decode();
      ...

**Comment by mgholam** [comment]
I'm sorry, but I don't understand what you are referring to, please make a clear example.

**Comment by BogdanovAnton** [comment]
Multithread work creates JSON like this
{
	"$type": "1",
	"ExtParamID": 7,
	"OrderArrange": 100,
	"CustomerID": 1,
	"Alignment": 0
}
it is a consequence of unplanned changes UsingGlobalTypes.
So
public deserializer(JSONParameters param)
{
this._params = param; // **You need cloning JSONParameters into new object here**
}
or to use only so that
var value = JSON.ToObject< TValue >(stringValue, **new JSONParameters()**);
var stringValue = JSON.ToJSON(value, **new JSONParameters()**);

**Comment by mgholam** [comment]
... and what is the problem with the json produced?

**Comment by BogdanovAnton** [comment]
**"$type": "1"** 
results

System.Exception: Cannot determine type
at fastJSON.deserializer.ParseDictionary(Dictionary`2 d, Dictionary`2 globaltypes, Type type, Object input)
at fastJSON.deserializer.ToObject(String json, Type type)
at fastJSON.deserializer.ToObject[T](String json)

**Comment by qaqz111** [comment]
I think I understood what @BogdanovAnton means.

When the `public static JSONParameters Parameters` of `static class JSON` is about to work with a **non-static deserializer object**, the very **non-static instance** of class deserializer which take the `Parameters` should get and save a copy from the static one. So that the instance will not be affected by the static JSONParameters Parameters' value changing during its working progress.

So, there should be a `MakeCopy()` of `JSONParameters`:
``` C#
    internal JSONParameters MakeCopy()
    {
        return new JSONParameters
        {
            DateTimeMilliseconds = DateTimeMilliseconds,
            EnableAnonymousTypes = EnableAnonymousTypes,
            IgnoreAttributes = new List<Type>(IgnoreAttributes),
            InlineCircularReferences = InlineCircularReferences,
            KVStyleStringDictionary = KVStyleStringDictionary,
            ParametricConstructorOverride = ParametricConstructorOverride,
            SerializeNullValues = SerializeNullValues,
            SerializerMaxDepth = SerializerMaxDepth,
            SerializeToLowerCaseNames = SerializeToLowerCaseNames,
            ShowReadOnlyProperties = ShowReadOnlyProperties,
            UseEscapedUnicode = UseEscapedUnicode,
            UseExtensions = UseExtensions,
            UseFastGuid = UseFastGuid,
            UseOptimizedDatasetSchema = UseOptimizedDatasetSchema,
            UseUTCDateTime = UseUTCDateTime,
            UseValuesOfEnums = UseValuesOfEnums,
            UsingGlobalTypes = UsingGlobalTypes,
        };
    }
```
And, `public static string ToJSON(object obj, JSONParameters param)` method should call it first:
``` C#
    public static string ToJSON(object obj, JSONParameters param)
    {
        param = param.MakeCopy();
        param.FixValues();
        ...
```
The constructor of `deserializer` should save a copy too:
``` C#
    public deserializer(JSONParameters param)
    {
        _params = param.MakeCopy();
    }
```

I looked into source of fastBinaryJSON and RaptorDB, the issue presence too.

**Comment by mgholam** [comment]
fixed in `2.1.31`
